### PR TITLE
Make our Prospector settings much stricter

### PR DESCRIPTION
**Pull request by seanh**

### CODE REVIEW

**Comment by seanh** [comment]
There are probably quite a lot of specific messages that we want to disable as well

**Comment by seanh** [comment]
Potential messages to ignore:

* pylint missing docstring
* pylint too few public methods
* pylint method could be a function (I actually like this one but it complains about things like test methods)
* Oh god I can't be bothered reading any more of this output!

**Comment by nickstenning** [comment]
Oops. #2234. Not a big deal, and I probably should have commented on #2232 as you did.

**Comment by nickstenning** [comment]
Yeah, that's a bit harsh.

**Comment by nickstenning** [comment]
Let's crank it up a bit at a time, rather than just telling the whole world via a badge that we are terrible people. (P.S. We're not.)

**Comment by seanh** [comment]
@nickstenning Can we go with this pull request then? What settings do you want to crank down?

**Comment by nickstenning** [comment]
Well, I think you should pull in the good-names setting, certainly. Many of those are just noise.

**Comment by seanh** [comment]
Waiting for landscape...

**Comment by seanh** [comment]
Vulture was just crashing for me anyway

**Comment by seanh** [comment]
A lot of this is still noise. I'm happy to go through it and disable messages (next week probably) and accept suggestions.

I think we should go for a middle-ground here. I have [a vim plugin](https://github.com/scrooloose/syntastic) that shows me prospector output in the sidebar as I code. The higher I have prospector cranked, the better the new code I write is going to be (though we should still disable messages we don't care about). Turning prospector up bit by bit may mean we're actually introducing more warnings as we write new code and just not seeing them cause we don't have it cranked high enough.

**Comment by nickstenning** [comment]
Agree on some of your points, but I think it's important that the SNR is reasonably high, otherwise we'll just ignore landscape, and that would be sad.

**Comment by seanh** [comment]
Alright. So let's see how high we can get it, by ignoring as much noise as possible? I assume landscape bot is gonna comment with another score at some point

**Comment by seanh** [comment]
That was 86% with strictness high, 83% with veryhigh. I think it's worth keeping it up. Anyone see any more messages worth disabling, either globally or case-by-case? There probably are some: <https://landscape.io/github/hypothesis/h/770/messages>. There are probably some particular problem files worth going through and fixing too: <https://landscape.io/github/hypothesis/h/770/modules>

Our biggest problem by far is invalid names.

**Comment by seanh** [comment]
If we get it up to I think 85% then we'll be Excellent instead of just Good

**Comment by seanh** [comment]
I want to disable pylint's protected-access warning, but only in tests

**Comment by tilgovi** [comment]
Would be happy to see it merge and I can look more once it's on master at what's worth fixing and disable things that profoundly annoy me. Not concerned about having an ~80% on our README.

**Comment by tilgovi** [comment]
Thanks for this.

**Comment by seanh** [comment]
Disabled protected-access in some test files (has to be done on a file-by-file basis afaict)

**Comment by seanh** [comment]
Rebased. Let's just wait and see what Travis and Landscape make of this in case the rebase broke something, but after that I'm happy with this.

**Comment by seanh** [comment]
I'm super happy to have this - I love having prospector on in my vim, and h's low prospector settings were kind of spoiling it. Just wish we had something like this for the rest of our codebase

**Comment by seanh** [comment]
We've pushed up into the _Excellent!_ rating :)

**Comment by tilgovi** [comment]
LGTM

**Comment by tilgovi** [comment]
Actually, there are a few of these I'd like to change. I'll comment in the diff.

File: .prospector.yaml [hunk 1]
```diff
@@ -1,14 +1,58 @@
...
+max-line-length: 79
 pep8:
-  options:
-    max-line-length: 79
+  full: true
 pylint:
   enable:
     - relative-import
+  disable:
+    - C0111  # Missing %s docstring
```
Comment (tilgovi) [inline .prospector.yaml]: Undecided about this one. We could fix so many of these in less than an hour of adding some docstrings.

File: .prospector.yaml [hunk 2]
```diff
@@ -1,14 +1,58 @@
...
+    - R0201  # Method could be a function
```
Comment (tilgovi) [inline .prospector.yaml]: I like this one because I like the non-OO push it provides.

File: .prospector.yaml [hunk 3]
```diff
@@ -1,14 +1,58 @@
...
+    - W0704  # Except doesn't do anything
```
Comment (tilgovi) [inline .prospector.yaml]: Definitely want this one enabled. We should not have any empty except clauses. At the very least we should usually be logging them.

File: .prospector.yaml [hunk 4]
```diff
@@ -1,14 +1,58 @@
...
+    # - k,v: dict iteration
+    # - db,fn: common abbreviations
+    # - fp: python idiom for file handles
+    #
+    # Some good "constant" names that pylint would otherwise reject:
+    #
+    # - log: common in "log = logging.getLogger(__name__)" pattern
+    # - parser: common in modules that use argparse
+    #
+    good-names: _,i,j,k,v,e,db,fn,fp,log,parser
```
Comment (tilgovi) [inline .prospector.yaml]: Nice.

File: .prospector.yaml [hunk 5]
```diff
@@ -1,14 +1,58 @@
...
+
+    # Allow longer function and method names.
```
Comment (tilgovi) [inline .prospector.yaml]: Could we handle this instead by saying '(test_)|(<the default definition>)'?

**Comment by tilgovi** [comment]
What's protected access about?

**Comment by tilgovi** [comment]
Finally, go have a weekend :)

**Comment by seanh** [comment]
It's accessing a function/method with an underscore, from another module/class. If you want to unit test that module/class you have to do it. (In Java the compiler wouldn't let you do this even in tests, you always had to test the public methods of classes only.)

File: .prospector.yaml (hunk 5 above)
Comment (seanh) [inline .prospector.yaml]: Good idea! That should work let's try it

File: .prospector.yaml (hunk 4 above)
Comment (seanh) [inline .prospector.yaml]: That was @nickstenning !

File: .prospector.yaml (hunk 3 above)
Comment (seanh) [inline .prospector.yaml]: Cool, let's put it back

File: .prospector.yaml (hunk 2 above)
Comment (seanh) [inline .prospector.yaml]: I like this one as well but it triggers very often in test methods. We could disable it in the test files specifically instead

File: .prospector.yaml (hunk 1 above)
Comment (seanh) [inline .prospector.yaml]: I'm happy for this to go back if you want

**Comment by seanh** [comment]
Watching a movie now :) Ciao

**Comment by tilgovi** [comment]
I sort of buy into the idea that testing protected methods and functions is not good.

If it's not exercisable from the public interface, should it be tested?

I think keeping that warning maybe encourages us to create better public interfaces and to test behavior rather than implementation.

**Comment by tilgovi** [comment]
I'm going to get this on master. Do more tweaks if you like from there.

**Pull request merged**